src/event/cloudevent.ts

```diff
@@ -189,11 +189,11 @@ See: https://github.com/cloudevents/spec/blob/v1.0/spec.md#type-system`);
    * @throws if the CloudEvent does not conform to the schema
    * @return {CloudEvent} returns a new CloudEvent
    */
-  public cloneWith(
-    options: CloudEventV1 | CloudEventV1Attributes | CloudEventV1OptionalAttributes,
-    strict = true,
-  ): CloudEvent {
-    return new CloudEvent(Object.assign({}, this.toJSON(), options) as CloudEvent, strict);
+  public cloneWith(options: Partial<Exclude<CloudEventV1<never>, "data">>, strict?: boolean): CloudEvent<T>;
+  // If the cloned event has a data property, the returned event should use the provided type
```

Comment:
I think you can also add a JSDoc style comment here and it will show up in VSCode autocompletion for this specific overload.